doc/syslog/syslog-rate-limit-design.md

```diff
@@ -0,0 +1,383 @@
+
+This document is the design document for syslog message rate limit configuration per container.
+
+### Definitions/Abbreviations
+
+N/A
+
+### Overview
+
+Logging in SONiC is organized with rsyslogd. Each container has its own rsyslogd instance plus a daemon running on host side. The rsyslogd instance which is running on the host is used to collect the messages from within containers and store them at certain path (E.g. /var/log/syslog). Rsyslog config file are generated from templates:
+
+- Container scope:
+  - Multi ASIC: https://github.com/Azure/sonic-buildimage/blob/master/files/image_config/rsyslog/rsyslog-container.conf.j2
+  - Single ASIC: https://github.com/Azure/sonic-buildimage/blob/master/dockers/docker-base/etc/rsyslog.conf
+- Host scope: https://github.com/Azure/sonic-buildimage/blob/master/files/image_config/rsyslog/rsyslog.conf.j2
+
+Currently, each container has hardcoded message rate limiting to avoid receiving flooded log messages:
+
+```
+$SystemLogRateLimitInterval 300
+$SystemLogRateLimitBurst 20000
+```
+There is no rate limiting configured on host side for now.
+
+The SystemLogRateLimitInterval determines the amount of time that is being measured for rate limiting. The SystemLogRateLimitBurst defines the amount of messages, that have to occur in the time limit of SystemLogRateLimitInterval, to trigger rate limiting. For example, SystemLogRateLimitInterval=300, SystemLogRateLimitBurst=20000, it means that if one daemon generate more than 20000 messages in 300 seconds, rsyslogd will start to drop messages after that(FIFO).
+
+This feature allows user to configure SystemLogRateLimitInterval and SystemLogRateLimitBurst for host, containers.
+
+### Requirements
+
+- Support syslog message rate limit configuration for host and containers
+- New CLI for rate limit configuration
+- Have rate limit configuration persistent
+- Default rate limit shall be applied if no configuration provided
+- Both multi ASIC and single ASIC platform shall use rsyslog-container.conf.j2 as template
+- APP extension shall be able to declare its capability of this feature
+- APP extension is responsible for reading/listening rate limit configuration from CONFIG DB and set rsyslog configuration accordingly.
+
+### Architecture Design
+
+1. CLI is only responsible for putting rsyslog configuration to CONFIG DB
+2. hostcfgd shall be extended to handle host side rsyslog configuration by listening CONFIG DB change
+3. A new daemon containercfgd shall be added to each container to handle container side rsyslog configuration
+4. rsyslog.conf.j2 and rsyslog-container.conf.j2 shall be extended to accept template variable for SystemLogRateLimitInterval and SystemLogRateLimitBurst.
+5. rsyslog.conf for single ASIC platform will be removed and replaced by rsyslog-container.conf.j2
+6. App extension shall handle rsyslog configuration itself
+
+#### CLI flow
+
+![cli_rate_limit_config_flow](/doc/syslog/images/cli_rate_limit_config_flow.svg)
+
+#### Host side flow
```

Comment:
@Junchao-Mellanox this approach introduces double host side configuration: first time with `rsyslog-config` service and second time with `hostcfgd`

Reply:
no, if configuration does not change, hostcfgd will not restart rsyslog

Reply:
@Junchao-Mellanox how do you plan to achieve that? The only reliable way which i see is to parse the `config_db.json` (taking into consideration also `init_cfg.json`) and fill daemon's internal cache on start

Reply:
We will parse /etc/rsyslog.conf for the first time, and comparing its configuration with CONFIG DB value.

Reply:
@Junchao-Mellanox what is the guarantee that `rsyslog.conf` will be up-to-date before the `hostcfgd` is started?

Reply:
Please check hostcfgd logic: https://github.com/sonic-net/sonic-host-services/blob/bc8698d1d760fefedaeb4742ad19b25ef2b3c17b/scripts/hostcfgd#L1661. Hostcfgd will delay itself by waiting for other system services. Thus, rsyslog-config.service and rsyslog.service should be started before hostcfgd,